src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -256,22 +250,30 @@ private Task WriteFormattedInt32Async(int value, CancellationToken cancellationT
                 // CRLF for end of headers.
                 await WriteTwoBytesAsync((byte)'\r', (byte)'\n', cancellationToken).ConfigureAwait(false);
 
-                // Write body, if any
+                // Flush out anything buffered from the request headers to finish sending them.
+                await FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
> The problem is, I don't know how to make this work while keeping the "headers and initial write in a single send" optimization.

Right, that's why I was suggesting special-casing the common cases we care to optimize, e.g. ByteArrayContent.  But, I'm fine just leaving the flush up to the HttpContent implementation; it's a bit difficult thinking through exactly where one does that flush, but it's also not a common case (someone will do it once and then others will use their work).